localization/ndt_scan_matcher/src/map_update_module.cpp

```diff
@@ -0,0 +1,303 @@
+    rclcpp::ServicesQoS().get_rmw_qos_profile(), map_callback_group_);
+
+  pcd_loader_client_ = node->create_client<autoware_map_msgs::srv::GetDifferentialPointCloudMap>(
+    "pcd_loader_service", rmw_qos_profile_services_default);
+  while (!pcd_loader_client_->wait_for_service(std::chrono::seconds(1)) && rclcpp::ok()) {
+    RCLCPP_INFO(
+      logger_,
+      "Waiting for pcd loader service. Check if the enable_differential_load in "
+      "pointcloud_map_loader is set `true`.");
+  }
+
+  double map_update_dt = 1.0;
+  auto period_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
+    std::chrono::duration<double>(map_update_dt));
+  map_update_timer_ = rclcpp::create_timer(
+    node, clock_, period_ns, std::bind(&MapUpdateModule::map_update_timer_callback, this),
+    map_callback_group_);
+}
+
+void MapUpdateModule::service_ndt_align(
+  const tier4_localization_msgs::srv::PoseWithCovarianceStamped::Request::SharedPtr req,
+  tier4_localization_msgs::srv::PoseWithCovarianceStamped::Response::SharedPtr res)
+{
+  // get TF from pose_frame to map_frame
+  auto TF_pose_to_map_ptr = std::make_shared<geometry_msgs::msg::TransformStamped>();
+  tf2_listener_module_->get_transform(
+    clock_->now(), map_frame_, req->pose_with_covariance.header.frame_id, TF_pose_to_map_ptr);
+
+  // transform pose_frame to map_frame
+  const auto mapTF_initial_pose_msg = transform(req->pose_with_covariance, *TF_pose_to_map_ptr);
+  update_map(mapTF_initial_pose_msg.pose.pose.position);
+
+  if (ndt_ptr_->getInputTarget() == nullptr) {
+    res->success = false;
+    RCLCPP_WARN(logger_, "No InputTarget");
+    return;
+  }
+
+  if (ndt_ptr_->getInputSource() == nullptr) {
+    res->success = false;
+    RCLCPP_WARN(logger_, "No InputSource");
+    return;
+  }
+
+  // mutex Map
+  std::lock_guard<std::mutex> lock(*ndt_ptr_mutex_);
+
+  (*state_ptr_)["state"] = "Aligning";
+  res->pose_with_covariance = align_using_monte_carlo(ndt_ptr_, mapTF_initial_pose_msg);
+  (*state_ptr_)["state"] = "Sleeping";
+  res->success = true;
+  res->pose_with_covariance.pose.covariance = req->pose_with_covariance.pose.covariance;
+
+  last_update_position_ = res->pose_with_covariance.pose.pose.position;
+}
+
+void MapUpdateModule::callback_ekf_odom(nav_msgs::msg::Odometry::ConstSharedPtr odom_ptr)
+{
+  current_position_ = odom_ptr->pose.pose.position;
+
+  if (last_update_position_ == std::nullopt) {
+    return;
+  }
+  double distance = norm_xy(current_position_.value(), last_update_position_.value());
+  if (distance + lidar_radius_ > dynamic_map_loading_map_radius_) {
+    RCLCPP_ERROR_STREAM_THROTTLE(logger_, *clock_, 1, "Dynamic map loading is not keeping up.");
+  }
+}
+
+void MapUpdateModule::map_update_timer_callback()
+{
+  if (current_position_ == std::nullopt) {
+    RCLCPP_ERROR_STREAM_THROTTLE(
+      logger_, *clock_, 1,
+      "Cannot find the reference position for map update. Please check if the EKF odometry is "
+      "provided to NDT.");
```

Comment:
Please check if this line is appropriate to write C++

Reply:
FYI: Its due to this pre-commit fix: https://github.com/kminoda/autoware.universe/commit/293176725d29b69b4401a5045ef804229d9be620

I could not find the name of this literal, but I confirmed that it works.

Reply:
FYI: Concatenation section in https://en.cppreference.com/w/cpp/language/string_literal.